zrml/global-disputes/src/lib.rs

```diff
@@ -0,0 +1,544 @@
+
+mod benchmarks;
+mod global_disputes_pallet_api;
+mod mock;
+mod tests;
+pub mod weights;
+
+pub use global_disputes_pallet_api::GlobalDisputesPalletApi;
+pub use pallet::*;
+
+#[frame_support::pallet]
+mod pallet {
+    use crate::{weights::WeightInfoZeitgeist, GlobalDisputesPalletApi};
+    use core::marker::PhantomData;
+    use frame_support::{
+        ensure, log,
+        pallet_prelude::{
+            Decode, DispatchError, DispatchResultWithPostInfo, Encode, MaxEncodedLen, OptionQuery,
+            StorageDoubleMap, StorageMap, TypeInfo, ValueQuery,
+        },
+        sp_runtime::{traits::StaticLookup, RuntimeDebug},
+        storage::child::KillStorageResult,
+        traits::{
+            Currency, ExistenceRequirement, Get, IsType, LockIdentifier, LockableCurrency,
+            WithdrawReasons,
+        },
+        Blake2_128Concat, BoundedVec, PalletId, Twox64Concat,
+    };
+    use frame_system::{ensure_signed, pallet_prelude::OriginFor};
+    use sp_runtime::traits::{AccountIdConversion, CheckedDiv, Saturating, Zero};
+    use sp_std::{vec, vec::Vec};
+    use zeitgeist_primitives::types::OutcomeReport;
+    use zrml_market_commons::MarketCommonsPalletApi;
+
+    pub(crate) type BalanceOf<T> = <<T as Config>::Currency as Currency<AccountIdOf<T>>>::Balance;
+    pub(crate) type MarketIdOf<T> =
+        <<T as Config>::MarketCommons as MarketCommonsPalletApi>::MarketId;
+    pub type AccountIdOf<T> = <T as frame_system::Config>::AccountId;
+
+    pub type OutcomeInfoOf<T> = OutcomeInfo<BalanceOf<T>, OwnerInfoOf<T>>;
+    pub type WinnerInfoOf<T> = WinnerInfo<BalanceOf<T>, OwnerInfoOf<T>>;
+    type AccountIdLookupOf<T> = <<T as frame_system::Config>::Lookup as StaticLookup>::Source;
+    type OwnerInfoOf<T> = BoundedVec<AccountIdOf<T>, <T as Config>::MaxOwners>;
+    type LockInfoOf<T> = LockInfo<MarketIdOf<T>, BalanceOf<T>>;
+
+    #[derive(Encode, Decode, Clone, Eq, PartialEq, RuntimeDebug, TypeInfo)]
+    pub struct LockInfo<MarketId, Balance>(pub Vec<(MarketId, Balance)>);
+
+    impl<MarketId, Balance> Default for LockInfo<MarketId, Balance> {
+        fn default() -> Self {
+            Self(Vec::new())
+        }
+    }
+
+    impl<MarketId, Balance> MaxEncodedLen for LockInfo<MarketId, Balance>
+    where
+        Balance: MaxEncodedLen,
+        MarketId: MaxEncodedLen,
+    {
+        fn max_encoded_len() -> usize {
+            MarketId::max_encoded_len().saturating_add(Balance::max_encoded_len())
+        }
+    }
+
+    #[derive(Debug, TypeInfo, Decode, Encode, MaxEncodedLen, Clone, PartialEq, Eq)]
+    pub struct OutcomeInfo<Balance, OwnerInfo> {
+        pub outcome_sum: Balance,
+        pub owners: OwnerInfo,
+    }
+
+    #[derive(TypeInfo, Decode, Encode, MaxEncodedLen, Clone, PartialEq, Eq)]
+    pub struct WinnerInfo<Balance, OwnerInfo> {
+        pub outcome: OutcomeReport,
+        pub vote_sum: Balance,
+        pub is_finished: bool,
+        pub owners: OwnerInfo,
+    }
```

Comment:
Should we move these structs to `primitives/global_dispute.rs` ?

Reply:
add51c6ca1d8a2bcb790aa9a2e8a611cff3688f3

Reply:
I think they should be kept in the crate. If I remember correctly the existing types from pallets in the primitive crate were added there because we had issues with circular dependencies. From my perspective this is a design error in many cases and should not occur. I am in favor of keeping the types in the pallet's crate. Probably it's enough to keep it all in one types module (without nesting), feel free to apply additional logical grouping within that types module though (see Rikiddo pallet for an example).

Reply:
b24d4a341966a8ccb8a60f080d23da8718244924